core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -437,62 +438,76 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
     }
+
+    // Figure out which nodes have any effective blacklisting on them.  This means all nodes that
+    // are explicitly blacklisted, plus those that have *any* executors blacklisted.
     val nodesForBlacklistedExecutors = offers.filter { offer =>
       execBlacklist.contains(offer.executorId)
     }.map(_.host).toSet.toSeq
-    val nodesToCheck = nodeBlacklist ++ nodesForBlacklistedExecutors
+    val nodesWithAnyBlacklisting = nodeBlacklist ++ nodesForBlacklistedExecutors
+    // Similarly, figure out which executors have any blacklisting.  This means all executors that
+    // are explicitly blacklisted, plus all executors on nodes that are blacklisted.
     val execsForBlacklistedNodes = offers.filter { offer =>
       nodeBlacklist.contains(offer.host)
     }.map(_.executorId).toSeq
-    val executorsToCheck = execBlacklist ++ execsForBlacklistedNodes
+    val executorsWithAnyBlacklisting = execBlacklist ++ execsForBlacklistedNodes
 
-    // Schedule a taskset, do a bit of basic sanity checking (that our test is operating the way its
-    // supposed to).
+    // Schedule a taskset, and make sure our test setup is correct -- we are able to schedule
+    // a task on all executors that aren't blacklisted (even the ones implicitly blacklisted by the
+    // node blacklist).
     val firstTaskAttempts = taskScheduler.resourceOffers(offers).flatten
-    assert(firstTaskAttempts.size === 3 - executorsToCheck.size)
-    assert(firstTaskAttempts.size < offers.size)
+    assert(firstTaskAttempts.size === 3 - executorsWithAnyBlacklisting.size)
 
     // Now check that we haven't made too many calls to any of the blacklist methods.
+    // We should be checking our node blacklist, but it should be within the bound we defined above.
     verify(stageToMockTaskSetBlacklist(0), atMost(maxBlacklistChecks))
       .isNodeBlacklistedForTaskSet(anyString())
-    nodesToCheck.foreach { node =>
+    // We shouldn't ever consult the per-task blacklist for the nodes that have been blacklisted
+    // for the entire taskset, since the taskset level blacklisting should prevent scheduling
+    // from ever looking at specific tasks.
+    nodesWithAnyBlacklisting.foreach { node =>
       verify(stageToMockTaskSetBlacklist(0), never)
         .isNodeBlacklistedForTask(meq(node), anyInt())
     }
-    executorsToCheck.foreach { exec =>
-      // If we had a node-blacklist, then we could tighten the next check to *never*.  But, it also
-      // doesn't particular matter if we the executor check happens in addition.
+    executorsWithAnyBlacklisting.foreach { exec =>
+      // We should be checking our executor blacklist, but it should be within the bound defined
+      // above.  Its possible that this will be significantly fewer calls, maybe even 0, if there
+      // is also a node-blacklist which takes effect first.  But this assert is all we need to
+      // avoid an O(numPendingTask) slowdown.
```

Comment:
Reading this comment and thinking about the randomness made me wonder if this test should be run in a loop where it gets run a bunch of times, to try to force more of the random orderings of offers? (this is assuming you don't like the suggestion to simplify the test to just have the blacklisted executors / nodes)